### PR TITLE
YAML config test coverage

### PR DESCRIPTION
**Pull request by buddhikac96**
This PR includes Unit Test coverage for YAML configuration parsing.

### CODE REVIEW

**Comment by NextTurn** [comment]
Don't have capital letters everywhere.

**Comment by NextTurn** [comment]
Again?

image

**Comment by buddhikac96** [comment]
> Again?
> 
> image

I don't know why it happens. :( Now I hard reset my master to upstream master. I think now it will be okay??

**Comment by NextTurn** [comment]
I believe these are all the steps to safely resolve this.

```sh
git rebase master yaml-test-coverage
git status
vi src/Core/ServiceWrapper/winsw.csproj
vi src/Test/winswTests/winswTests.csproj
git add .
git rebase --continue
git rebase --skip # if Git tells you to do so
git push origin yaml-test-coverage --force
```

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs [hunk 1]
```diff
@@ -277,12 +284,21 @@ public override string ErrFilePattern
...
                 {
-                    if (this.ZipOlderThanNumDaysYamlLog != null)
+                    int? zipolderthannumdays = null;
+
+                    if (!string.IsNullOrEmpty(this.ZipOlderThanNumDaysYamlLog))
                     {
-                        return this.ZipOlderThanNumDaysYamlLog;
+                        if (!int.TryParse(this.ZipOlderThanNumDaysYamlLog, out int zipolderthannumdaysValue))
+                        {
+#pragma warning disable S2372 // Exceptions should not be thrown from property getters
```
Comment (NextTurn) [inline src/Core/WinSWCore/Configuration/YamlConfiguration.cs:293]: Without the analyzer installed in this repo these directives should not be committed.

Also, please remove those which has been merged into the code base.

File: src/Test/winswTests/ServiceDescriptorYamlTest.cs [hunk 2]
```diff
@@ -1,149 +1,471 @@
...
+        private const string Username = "User";
+        private const string Password = "Password";
+        private const string Domain = "Domain";
+        private const string AllowServiceAccountLogonRight = "true";
+
+
+        [SetUp]
+        public void SetUp()
+        {
+            string yaml = $@"id: service.exe
```
Comment (NextTurn) [inline src/Test/winswTests/ServiceDescriptorYamlTest.cs:24]: Starting the string on the next line increases readability.

File: src/Test/winswTests/ServiceDescriptorYamlTest.cs [hunk 3]
```diff
@@ -1,149 +1,471 @@
...
-            var configs = ServiceDescriptorYaml.FromYaml(MinimalYaml).Configurations;
+            string yaml = $@"id: service.exe
+name: Service
+description: The Service.
+executable: node.exe
+arguments: My Arguments
+startMode: manual
+log:
+    mode: roll
+    logpath: c:\logs
```
Comment (NextTurn) [inline src/Test/winswTests/ServiceDescriptorYamlTest.cs:82]: It doesn't make sense to include unrelated setting entries.

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 1 above)
Comment (buddhikac96) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:293]: Make sense. But when remove all #pragma warnings, build is failing. How to resolve this?

Comment (NextTurn) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:293]: Tweak your own build environment.

File: src/Core/ServiceWrapper/winsw.csproj [hunk 4]
```diff
@@ -20,6 +20,10 @@
     <ILMergeVersion>3.0.41</ILMergeVersion>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (NextTurn) [inline src/Core/ServiceWrapper/winsw.csproj:23]: You added back all removed items when "resolving conflicts".

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 1 above)
Comment (NextTurn) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:293]: This is unrelated to Visual Studio or .NET SDK.

**Review by oleg-nenashev — APPROVED** [review]
I tentatively approve this PR, but I agree with @NextTurn that the '#pragma warning disable S2372 ' comment should be ideally addressed. Not blocking with assumption we clean it up for v3 during the upmerge

**Review by NextTurn — APPROVED** [review]
I guess @buddhikac96 hasn't figured out what's going wrong with his build environment. Let's merge this since there are no other outstanding review comments.

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 1 above)
Comment (oleg-nenashev) [inline src/Core/WinSWCore/Configuration/YamlConfiguration.cs:293]:
```suggestion
//FIXME: Remove the warning from the codebase
#pragma warning disable S2372 // Exceptions should not be thrown from property getters
```

**Comment by oleg-nenashev** [comment]
Added a warning for the concerning pragma

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs (hunk 1 above)
Comment (NextTurn) [reply src/Core/WinSWCore/Configuration/YamlConfiguration.cs:294]:
```
Single line comment should begin with a space.
```

**Comment by oleg-nenashev** [comment]
> Configuration\YamlConfiguration.cs(293,1): error SA1005: Single line comment should begin with a space. 

🤷

File: src/Core/WinSWCore/Configuration/YamlConfiguration.cs [hunk 5]
```diff
@@ -277,12 +284,22 @@ public override string ErrFilePattern
...
                 {
-                    if (this.ZipOlderThanNumDaysYamlLog != null)
+                    int? zipolderthannumdays = null;
+
+                    if (!string.IsNullOrEmpty(this.ZipOlderThanNumDaysYamlLog))
                     {
-                        return this.ZipOlderThanNumDaysYamlLog;
+                        if (!int.TryParse(this.ZipOlderThanNumDaysYamlLog, out int zipolderthannumdaysValue))
+                        {
+//FIXME: Remove the build env specific warning suppression from the codebase
```
Comment (oleg-nenashev) [inline src/Core/WinSWCore/Configuration/YamlConfiguration.cs:293]:
```suggestion
// FIXME: Remove the build env specific warning suppression from the codebase
```

**Pull request merged**